File: Java/LeetCode/src/PlayWithAlgorithms/MergeSort.java
package PlayWithAlgorithms;

/**
 * Created by Lix on 2017/8/17.
 */
public class MergeSort {

    private MergeSort() {}

    public static void sort(Comparable[] arr) {
        sort(arr, 0, arr.length-1);
    }

    private static void sort(Comparable[] arr, int l, int r) {
        if (l >= r) {
            return;
        }

        int mid = (l + r) / 2;
        sort(arr, l, mid);
        sort(arr, mid+1, r);
        if (arr[mid].compareTo( arr[mid+1] ) > 0) {
            merge(arr, l, mid, r);
        }
    }

    private static void merge(Comparable[] arr, int l, int mid, int r) {
        Comparable[] aux = new Comparable[r-l+1];
        for (int i = l; i <= r; i++) {
            aux[i-l] = arr[i];
        }

        int i = l, j = mid + 1;
        for (int k = l; k <= r; k++) {
            if (i > mid) {
                arr[k] = aux[j-l];
                j++;
            } else if (j > r) {
                arr[k] = aux[i - l];
                i++;
            } else if (aux[i-l].compareTo( aux[j-l] ) < 0) {
                arr[k] = aux[i-l];
                i++;
            } else {
                arr[k] = aux[j-l];
                j++;
            }
        }
    }
}

File: src/algs4/sort/merge-sort.ts
import BaseSort from './base-sort'

/**
 * 归并排序
 */
export default class MergeSort<T> extends BaseSort<T> {
  private aux: T[] // 归并所需的辅助数组

  sort(arr: T[]) {
    this.aux = []
    this._sort(arr, 0, arr.length - 1)
  }

  private _sort(arr: T[], lo: number, hi: number) {
    if (hi <= lo) return
    const mid = lo + Math.floor((hi - lo) / 2)
    this._sort(arr, lo, mid) // 将左半边排序
    this._sort(arr, mid + 1, hi) // 将右半边排序
    this.merge(arr, lo, mid, hi) // 归并结果
  }

  private merge(arr: T[], lo: number, mid: number, hi: number) {
    // 将 arr[lo..mid] 和 arr[mid+1..hi] 归并jj
    let i = lo
    let j = mid + 1
    for (let k = lo; k <= hi; k++) {
      // 将 arr[lo..hi] 复制到aux[lo..hi]
      this.aux[k] = arr[k]
    }
    for (let k = lo; k <= hi; k++) {
      // 归并回到arr[lo..hi]
      if (i > mid) arr[k] = this.aux[j++]
      else if (j > hi) arr[k] = this.aux[i++]
      else if (this.less(this.aux[j], this.aux[i])) arr[k] = this.aux[j++]
      else arr[k] = this.aux[i++]
    }
  }
}
